# Prerequisitos con Python - Parte 3

En esta sección veremos las diferentes maneras que ofrece python para control de flujo del algoritmo, tanto python como otros lenguajes de programación manejan la manera de que se decide ejecutar en el algoritmo pero con la sintaxis de programación especificad para cada lenguaje.

Aprenderá a hacer estructuras condicionales con código para bifurcar haciendo:
- condicionales
- for y while
- break y continue
- usar funciones especiales como zip y enumerate
- listas de comprensión

#### Condicionales

Los códigos se ejecutan secuencialmente, de arriba hacia abajo.  Sin embargo existen situaciones que necesitamos que solamente una sección de código se ejecute en específico.

Por ejemplo, digamos que su aplicación hace cobro solamente los 15 de cada mes, sin embargo siempre está activa todo el tiempo y validando que lo pactado cuando se de sea ejecutado.

Debajo mostramos el posible código de esto...

In [1]:
dia = 15
if dia == 15:
    print('Ejecutar cobro')
    
dia = 2
if dia == 15:
    print('Ejecutar cobro')

Ejecutar cobro


Una sentencia if se ejecuta dependiendo de la condicion interna de evaluación.

    if <condicion>:
        <codigo>

Donde la condicion a evaluar da un resultado booleano y si es evaluada como cierta (True) hace pase para ejecutar el código identado.

#### Ahora ud

Haga un flujo de código donde permita evaluar si puede aplicar a una ubicación dependiendo de su ingreso

- 1-500     -- Pacora
- 500-1000  -- Las Cumbres
- 1000-3000 -- El Dorado
- 3000-5000 -- San Francisco
- 5000      -- Altos del Maria

In [1]:
i= 1050
if i<=500:
    print('Pacora')
if 500<i<=1000:
    print('Las Cumbres')
if 1000<i<=3000:
    print('El Dorado')
if 3000<i<=5000:
    print('San Francisco')
if 5000<i:
    print('Altos del Maria')

El Dorado


#### Adivina el número

Provea la estructura condicional donde pueda evaluar si un numero es mayor, menor o ha sido acertado


In [2]:
respuesta = 18
numero    = int(input())

if respuesta<numero:
    resp = 'Te fuiste muy arriba'
if respuesta>numero:
    resp = 'Te fuiste muy abajo'
if respuesta==numero:
    resp = 'Ese es el numero, atinaste'
print('Numero escogido:', numero)    
print(resp)

1
Numero escogido: 1
Te fuiste muy abajo


#### Pago de impuestos

Haga un programa donde si gana mas de 900 dolares calcule e imprima el seguro social y educativo a descontar.
Nota:  
- Tome 9.75% como seguro social
- Tome 1.25% como seguro educativo

In [4]:
S=int(input())
print('Salario:', S)
if S<=900:
    print('No Hay Descuento.')
if S>900:
    SS=S*0.0975
    SE=S*0.0125
    ST= S-(SS+SE)
    print('Salario Total:', ST)


1200
Salario: 1200
Salario Total: 1068.0


### Expresiones boolanas para estructuras condicionales

Todo lo visto anteriormente solamente evalua una condicion de experesion, pero se pueden insertar más de una expresión en la sentencia if.

Podemos también insertar algebra para hacer calculos antes de toma de decisiones

In [4]:
edad = 24
if 18 < edad <= 40:
    print('Usted es un adulto joven')

Usted es un adulto joven


In [8]:
# w/h es la razon de aspecto de una persona
w = 0.8
h = 1.7
if 2.5 > h/w > 1.5:
    print('Ud es normal y no sufre de gigantismo')

Ud es normal y no sufre de gigantismo


Tambien podemos evaluar las condiciones utilizando algebra booleana comun y corriente con expresiones como 'and' 'or' 'not'.  Para el caso siguiente note que necesitamos que ambas variables sean ciertas para ejecutar esta linea de código.

In [9]:
haveCar = True
haveGas = True
if haveCar and haveGas:
    print('Puedo pasear en auto')

Puedo pasear en auto


Compliquemos un poco más el asunto viendo una estructura mas compleja.

Asumamos que hemos sido evaluados para ingresar a una empresa bajo las siguientes condicionales.

In [12]:
haveSkills = True
knowPython = True
partTime   = False
LiveIn     = 'PANAMA'
problematic = False
teamwork   = 0.7
if haveSkills and knowPython and (LiveIn == 'PANAMA' or LiveIn == 'COLON') and (not problematic and teamwork > 0.65):
    print('El candidato encaja en el perfil')

El candidato encaja en el perfil


En resumen, debemos tener en cuenta que para que la condición a evaluar dentro del if se ejecute el total de las expresiones debe ser booleanamente como cierta.

#### AHora ud

Imagine que un robot va a detectar una colision y la metrica es distancia.  Genre una parte de expresion booleana que determine si colisiona cuando la distancia decae por debajo de 0.5 metros.

In [5]:
D=float(input())
B= True
print('Distancia detectada: ', D)
if B and D < 0.5:
    print('Peligro')


0.4
Distancia detectada:  0.4
Peligro


Cuáles seran los resultados de estas condiciones:

- tiempo = 10
- distancia = 0.5
- velocidad = 0.2

Evaluar para
- `10 > tiempo > 1`
- `velocidad*tiempo > 4 and distancia > 0.2 and tiempo`
- `not (velocidad/t**2 > 0.1) or distancia < 1`
- `velocidad*tiempo > 0.5 or distancia/tiempo > 10`

In [6]:
T=10
D=0.5
V=0.2
if 10>T>1:
    print('Resultado A')
if(V*T > 4) and (D > 0.2 and T):
    print('Resultado B')
if not (V/T**2 > 0.1) or (D < 1):
    print('Resultado C')
if (V*T > 0.5) or (D/T > 10):
    print('Resultado D')

Resultado C
Resultado D


### Ciclos For

Ahora que sabe utilizar instrucciones condicionales podemos ver los ciclos, veremos primero ciclos for y luego ciclos while que son las estructuras cíclicas que maneja este lenguaje.

Hablemos del objeto `iterable`.  Este objeto puede retornar más de un elemento a la vez.  Pueden retornar cualesquiera de estos 3 elementos `listas, cadenas o tuples` ademas de `diccionarios y archivos`.

Veamos un ejemplo...

In [50]:
algunas_provincias = ['PanAma', 'ColOn', 'DArien', 'VerAguas', 'ChIriqui', 'HErrera', 'LOs SAntOs']

for provincia in algunas_provincias:
    print(provincia.upper())

PANAMA
COLON
DARIEN
VERAGUAS
CHIRIQUI
HERRERA
LOS SANTOS


¿Que sucede?.  Del ejemplo anterior tenemos que disectar un poco.
- For nos dice que es un ciclo o un bucle repetitivo
- algunas_provincias  es el iterable o la variable que contiene los elementos en donde vamos a ciclar
- provincia    es la variable que contiene el estado actual de ciclo que esta procesando

En la primera iteración será Panamá, luego colon y así sucesivamente hasta llegar al final de la lista, la variable provincia irá tomando ese valor.

Podemos también crear lazos para ingresar una nueva lista o elementos a la misma lista.  Veamos como con el mismo ejemplo podemso hacerlo.

In [17]:
algunas_provincias = ['PanAma', 'ColOn', 'DArien', 'VerAguas', 'ChIriqui', 'HErrera', 'LOs SAntOs']
en_mayusculas = []
for provincia in algunas_provincias:
    en_mayusculas.append(provincia.upper())

print(en_mayusculas)

['PANAMA', 'COLON', 'DARIEN', 'VERAGUAS', 'CHIRIQUI', 'HERRERA', 'LOS SANTOS']


Bien, hemos visto como insertar en una lista, pero que pasa si queremos modificar. Podemos utilizar la función range para esto.
- range(start, stop, step)
    - Nota:  Los argumentos todos son números enteros para este caso
    - start es donde inicia la secuencia
    - stop es donde se detiene o la condicion final
    - step es de cuanto en cuanto va avanzando el ciclo
    - Si no se especifican start o step son 0 y 1 respectivamente
    
Entendamos un poco como opera range antes de seguir

In [20]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Observo que cuando se especifica range() este empieza de 0 y termina en n-1, es por esto que la secuencia es de 0 a 9 si se especifica un valor final de 10.

In [21]:
list(range(4, 9))

[4, 5, 6, 7, 8]

In [22]:
list(range(1,2))

[1]

Aquí entonces notara que solamente hace uso del valor start = 4 y stop = 9, que en la lista es el 8 (o 9-1).  Igualmente para start = 1 y stop = 2 note que solamente ejecuta un solo paso.

Note que no utilizamos print para imprimir el contenido de range() sino una lista, esto es porque si utilizamos print nos da el tipo de datos.  La manera de hacerlo correctamente es utilizando lo anterior o también con ciclos for.

In [23]:
print(range(10))

for n in range(10):
    print(n)

range(0, 10)
0
1
2
3
4
5
6
7
8
9


Otra manera de acceder cada elemento de una lista es por medio de su índice, veamos como se hace esto con un ciclo for

In [29]:
algunas_provincias = ['PanAma', 'ColOn', 'DArien', 'VerAguas', 'ChIriqui', 'HErrera', 'LOs SAntOs']
for provincia_num in range(len(algunas_provincias)):
    algunas_provincias[provincia_num] = algunas_provincias[provincia_num].lower()
    
print(algunas_provincias)

['panama', 'colon', 'darien', 'veraguas', 'chiriqui', 'herrera', 'los santos']


#### Ahora practique ud

1 - Use ciclos for para listar los elementos de la tabla periódica correctamente (primera letra en mayúscula).  Pista: use la funcion title().

In [7]:
table_periodica = ['na', 'br', 'ca', 'cu', 'au', 'nd', 'ag']
for elementos in range(len(table_periodica)):
    table_periodica[elementos]= table_periodica[elementos].title()
print(' , '.join(table_periodica))

Na , Br , Ca , Cu , Au , Nd , Ag


Respuesta esperada:

['Na', 'Br', 'Ca', 'Cu', 'Au', 'Nd', 'Ag']

2 - Escriba con ciclos for el siguiente trabalenguas con espacios.

In [8]:
trabalenguas = ['Por', 'desenredar', 'el', 'enredo', 'que', 'ayer', 'enredé.', 'Hoy', 'enredo',
                'el', 'desenredo', 'que', 'desenredé', 'ayer.']

" ".join([palabras for palabras in trabalenguas])


'Por desenredar el enredo que ayer enredé. Hoy enredo el desenredo que desenredé ayer.'

Respuesta esperada:

'Por desenredar el enredo que ayer enredé. Hoy enredo el desenredo que desenredé ayer'

3 - Escriba un programa que calcule el factorial del número 6 solamente teniendo como ingreso el número 6

In [10]:
N=6
F=1
for i in range(1, N +1):
    F= F*i
print('El factorial es:', F)

El factorial es: 720


Respuesta esperada:

720

Haga la progresión de sumas del número 30, es decir:
- 1
- 1+2
- 1+2+3
- 1+2+3...+30

In [11]:
I=1
F=30
R=0
for i in range(I ,F+1):
    R +=i
print(R)

465


Resultado esperado:

465

Escriba los siguientes nombres en minuscula separados por '-'

- Himura Kenshin
- Kamiya Kaoru
- Myōjin Yahiko
- Sagara Sanosuke
- Takani Megumi
- Shinomori Aoshi
- Saitō Hajime
- Makimachi Misao

In [12]:
Nombres=['Himura Kenshin', 'Kamiya Kaoru', 'Myōjin Yahiko', 'Sagara Sanosuke', 'Takani Megumi', 'Shinomori Aoshi', 'Saitō Hajime', 'Makimachi Misao']
A=None
Nombresf=[]
for n in Nombres:
    A=str(n)
    A1= A.replace(' ', '-')
    Nombresf.append(A1)
Nombresf='\n '.join(Nombresf)
print(Nombresf.lower())


himura-kenshin
 kamiya-kaoru
 myōjin-yahiko
 sagara-sanosuke
 takani-megumi
 shinomori-aoshi
 saitō-hajime
 makimachi-misao


Respuesta esperada:

- himura-kenshin
- kamiya-kaoru
- myōjin-yahiko
- sagara-sanosuke
- takani-megumi
- shinomori-aoshi
- saitō-hajime
- makimachi-misao

### Iteracion con diccionarios

Con diccionarios la situación cambia un poco, recordemos que un diccionario tiene 'keys' y 'values'.  Veamos que pasa si interactuamos con uno.

In [39]:
numeros = {0:'cero', 1:'uno', 2:'dos', 3:'tres', 4:'cuatro', 5:'cinco',
           6:'seis', 7:'siete', 8:'ocho', 9:'nueve', 10:'diez'}

for num in numeros:
    print(num)

0
1
2
3
4
5
6
7
8
9
10


Notamos que si tendemos a imprimir de una lista entonces solamente tenemos las llaves (keys), que pasa si queremos los valores, pues entonces tenemos que acceder la funcion items().

In [40]:
for key, val in numeros.items():
    print(key, val)

0 cero
1 uno
2 dos
3 tres
4 cuatro
5 cinco
6 seis
7 siete
8 ocho
9 nueve
10 diez


igualmente, podemos acceder partes del diccionario con keys() o values() respectivamente

In [44]:
for key in numeros.keys():
    print(key)

0
1
2
3
4
5
6
7
8
9
10


In [45]:
for val in numeros.values():
    print(val)

cero
uno
dos
tres
cuatro
cinco
seis
siete
ocho
nueve
diez


#### Ahora intentelo ud

Intente sumar todas las edades en el diccionario

In [13]:
edades = {'Juan':10, 'Javier': 4, 'Jonas':8, 'Nuria': 20, 'Edith': 18, 'Ana':14}
resultado=0
for i in edades.values():
    resultado +=i
print(resultado)

74


Respuesta esperada:

74

###  Ciclos While

Los ciclos for se ejecutan dependiendo del control deseado, es decir podemos ejecutarla por medio de un limitador de condición final.  Ejecuta entonces en un rango de valores o también por medio del contenido de una lista.

Considere el siguiente ejemplo de un ciclo while, que es otra estructura de repeticion en python.

In [66]:
numeros = [1, 4, 9, 10, 120, 22, 43]
reversal = []
i = 0
while len(numeros):
    reversal.append(numeros.pop())
    i +=1
    
print(reversal)

[43, 22, 120, 10, 9, 4, 1]


### Ahora ud

Presente el resultado del factorial de 6 pero ahora utilizando ciclos while



In [14]:
f=6
r=1
while(f>=1):
    r=r*f
    f=f-1
print(r)

720


Respuesta esperada:
    
720

De la siguiente lista solamente liste los numeros que son impares utilizando ciclo while

In [15]:
lista = list(range(16))
impar=1
while (impar<len(lista)):
    if lista[impar]%2!=0:
       print('\n',lista[impar], end='')
    
    impar +=1
 




 1
 3
 5
 7
 9
 11
 13
 15

Respuesta esperada:

Una lista en orden ascendente o descendente como sigue:

    15 
    13
    11 
    9
    7
    5
    3
    1
    
    ó
    
    1
    3
    5
    7
    9
    11
    13
    15

Encuentre potencia al cuadrado mas cercana al número que se da a continuación utilizando el ciclo while.

Nota:  Solo utilice valores enteros, no decimales

In [16]:
num = 456
R=0
while (R+1)**2<num:
    R+=1
    
print('El número más cercano es {}'.format(R))


El número más cercano es 21


Respuesta esperada:

El número más cercano es 21

### Break y Continue

Como vimos anteriormente los ciclos for y while hacen lazos.  El ciclo for especificamente culmina hasta en contrar el final de una secuencia y el ciclo while hasta encontrar una condicion final.

En la mayoria de los casos esto es suficiente, pero a veces necesitamos más control del flujo final de estos lazos.  

Para el caso de break termina el ciclo y se da luego de que una condición se alcanza.

En el caso de continue, al alcanzar la condicion retornamos a realizar la iteraccion para el siguiente dato del ciclo.

Esto se verá mejor en un ejemplo.


Para el siguiente ejemplo suponga que ud tiene un contenedor de 40 pies y que la carga máxima es 3750 kg.  

Vamos a realizar un programa que nos haga cargar todos los elementos que sean menors a 500 kg.

In [101]:
# item, peso (kg)
items = [('frutas', 320), ('vegetales', 655), ('jugetes', 880), ('cereales', 375), 
         ('detergente', 450), ('hogar', 200), ('farmacia', 499), ('papel', 1000), ('otros', 932)]

contenedor = []
total_weight = 0
for item in items:
    name, weight = item
    if weight > 500:
        continue # no cargar al contenedor
    else:
        total_weight += weight              
        contenedor.append(name)

print(f'Cargamos al contenedor {contenedor} y el peso total de la mercancia es {total_weight} kg')

Cargamos al contenedor ['frutas', 'cereales', 'detergente', 'hogar', 'farmacia'] y el peso total de la mercancia es 1844 kg


El mismo contenedor de 40 pies y de carga máxima es 3750 kg.  

Vamos a realizar un programa que cuando detecte que la capacidad máxima sobrepasa los 3750 kg se detenga.


In [102]:
contenedor = []
total_weight = 0
for item in items:
    name, weight = item
    total_weight += weight        
    if total_weight >= 3750:
        total_weight -= weight
        break # no cargar más de la capacidad máxima
    else:              
        contenedor.append(name)

print(f'Cargamos al contenedor {contenedor} y el peso total de la mercancia es {total_weight} kg')

Cargamos al contenedor ['frutas', 'vegetales', 'jugetes', 'cereales', 'detergente', 'hogar', 'farmacia'] y el peso total de la mercancia es 3379 kg


#### Debugging

Una manera simple de hacer debug (depurar el código o encontrar errores) es utilizando print() como hemos visto... podemos utilizar los mismos en la hoja y ver que está sucediendo

In [108]:
contenedor = []
total_weight = 0
for item in items:
    name, weight = item
    total_weight += weight        
    if total_weight >= 3750:
        total_weight -= weight
        print(f'\n**carga total al final del ciclo {total_weight}**')        
        break # no cargar más de la capacidad máxima
    else:  
        print(f'cargando {name}, peso actual {total_weight}')
        contenedor.append(name)

print(f'\nCargamos al contenedor {contenedor} y el peso total de la mercancia es {total_weight} kg')

cargando frutas, peso actual 320
cargando vegetales, peso actual 975
cargando jugetes, peso actual 1855
cargando cereales, peso actual 2230
cargando detergente, peso actual 2680
cargando hogar, peso actual 2880
cargando farmacia, peso actual 3379

**carga total al final del ciclo 3379**

Cargamos al contenedor ['frutas', 'vegetales', 'jugetes', 'cereales', 'detergente', 'hogar', 'farmacia'] y el peso total de la mercancia es 3379 kg


### Mas de Continue en ciclos

A veces necesitamos esquivar solamente el primer valor o un solo valor, para esto podemos utilizar 'in' que vimos anteriormente en otros apartados.

Para el siguiente ejemplo vamos a pasar y eli

In [118]:
ufc_champions = ['Stipe Miocic', 'Israel Adesanya', 'Kamaru Usman', 'Khabib Nurmagomedov',
                 'Justin Gaethje', 'Alexander Volkanovski', 'Petr Yan', 'Deiveson Figueiredo']

ufc_champions_short = ['Israel Adesanya', 'Khabib Nurmagomedov', 'Justin Gaethje', 'Deiveson Figueiredo']

count = 0
for champ in ufc_champions:
    if champ not in ufc_champions_short:
        print('No fue encontrado este campeon')
        continue
    else:
        count = count + 1
        print(f'Contado el campeon {champ}')
        
print('Total de campeones encontrados ->', count)

No fue encontrado este campeon
Contado el campeon Israel Adesanya
No fue encontrado este campeon
Contado el campeon Khabib Nurmagomedov
Contado el campeon Justin Gaethje
No fue encontrado este campeon
No fue encontrado este campeon
Contado el campeon Deiveson Figueiredo
Total de campeones encontrados -> 4


#### Intentelo ud ahora

Generador de piropos.  

De la siguiente lista generará piropos pero excluyendo una lista de palabras ofensivas que también están en esta lista

In [17]:
palabras = ['Cretino', 'Vida mía', 'Baboso', 'Bocachancla','Cielo', 'Cielito', 'Mi vida', 'Mi luz', 
            'Luz de mis ojos', 'Bebecita', 'Cariño', 'Papacito', 'Zoquete', 'Cariño mío', 'Patan', 'Cari', 
            'Corazón', 'Mi rey', 'Lerdo','Osito', 'Bombón', 'Charlatan', 'Dulzura', 'Galletita', 
            'Fanfarrón']
ofensas = ['Baboso', 'Bocachancla', 'Charlatan', 'Cretino', 'Fanfarrón', 'Lerdo', 'Patan', 'Zoquete']
piropos=[]
for piropo in palabras:
    if piropo in ofensas:
        continue
    else:
        piropos.append(piropo)
piroposf=str(piropos)[1:-1] 
print(piroposf)



'Vida mía', 'Cielo', 'Cielito', 'Mi vida', 'Mi luz', 'Luz de mis ojos', 'Bebecita', 'Cariño', 'Papacito', 'Cariño mío', 'Cari', 'Corazón', 'Mi rey', 'Osito', 'Bombón', 'Dulzura', 'Galletita'


Respuesta esperada: 

Una lista de posibles piropos, p.e.

'Vida Mia Cielo', 'Vida Mia Cielito', ... 'Bombon Dulzura'

No se espera:
'Papacito Zoquete'

#### Zip y Enumerate

De ejemplos anteriores observamos una lista que poseia tuples

    items = [('frutas', 320), ('vegetales', 655), ('jugetes', 880), ('cereales', 375), 
         ('detergente', 450), ('hogar', 200), ('farmacia', 499), ('papel', 1000), ('otros', 932)]
         
Digamos que empezamos de dos listas, una de items y otra de pesos.

Se puede hacer una mejor manipulación de los datos si se almacenan de forma más eficiente, es para esto que utilizamos zip.

Zip, retorna un iterador que combina multiples iterables en una secuencia de tubples.  Cada tuple contiene elementos en la posicion de los iterables.

Veamos el siguiente ejemplo

In [120]:
item = ['frutas', 'vegetales', 'juguetes', 'cereales', 'detergente', 'hogar', 'farmacia', 'papel', 'otros']
peso = [320, 655, 880, 375, 450, 200, 499, 1000, 932]


print(list(zip(item, peso)))

[('frutas', 320), ('vegetales', 655), ('juguetes', 880), ('cereales', 375), ('detergente', 450), ('hogar', 200), ('farmacia', 499), ('papel', 1000), ('otros', 932)]


Tambien, como son de forma tuple, podemos desempaquetar el tuple como hicimos anteriormente, veamos como podemos hacer esto

In [121]:
#same as below comments
#elementos = zip(item, peso)
#for item, peso in elementos: # same as below
for n, w in zip(item, peso):
    print(n, w)

frutas 320
vegetales 655
juguetes 880
cereales 375
detergente 450
hogar 200
farmacia 499
papel 1000
otros 932


Tambien podemos aceder y hacer unzip de los elementos del zip

In [122]:
elementos = zip(item, peso)
item, peso = zip(*elementos)

print(item)
print(peso)

('frutas', 'vegetales', 'juguetes', 'cereales', 'detergente', 'hogar', 'farmacia', 'papel', 'otros')
(320, 655, 880, 375, 450, 200, 499, 1000, 932)


En ocasiones también nos interesa no solo descomprimir los valores de una lista sino también tener los indices de ella, esto se puede hacer con zip de la siguiente manera.

In [126]:
for i, elemento in zip(range(len(items)), item):
    print(i, elemento)

0 frutas
1 vegetales
2 juguetes
3 cereales
4 detergente
5 hogar
6 farmacia
7 papel
8 otros


Esto quizá es muy complicado, para hacer lo mismo podemos utilizar 'enumerate'.  Este nos dá el índice por medio de acceso de los elementos.  Este es el método preferido porque es más 'pythonic'.

In [127]:
for i, elemento in enumerate(item):
    print(i, elemento)

0 frutas
1 vegetales
2 juguetes
3 cereales
4 detergente
5 hogar
6 farmacia
7 papel
8 otros


#### Ahora intentelo ud

Genere un diccionario que contenga una letra de y coordenadas en el espacio utilizando zip, por ejemplo 'A':[12, 1.2, 13]
se dará una lista de etiquetas, y otra de coordenadas.

In [18]:
x0 = [0, 1, -2, 14, -5, 4, 5]
x1 = [1, 3, 5, 7, 9, 11, 13]
x2 = [2, 4, 6, 8, 10, 12, 14]
letras = ['A', 'B', 'C', 'D', 'F', 'E', 'F']
C= zip(x0,x1,x2)
DICC1= dict(zip(letras, C))
print(DICC1)

{'A': (0, 1, 2), 'B': (1, 3, 4), 'C': (-2, 5, 6), 'D': (14, 7, 8), 'F': (5, 13, 14), 'E': (4, 11, 12)}


Resultado esperado:

{'A': [0, 1, 2], 'B': [1, 3, 4], 'C': [-2, 5, 6], 'D': [14, 7, 8], 'F': [5, 13, 14], 'E': [4, 11, 12]}

Haga unizp de la lista inferior

In [19]:
tup = ((1,-1), (2,-2), (3,-3), (4,-4), (5,-5))
pos,neg= zip(*tup)

print(pos)
print(neg)

(1, 2, 3, 4, 5)
(-1, -2, -3, -4, -5)


Respuesta esperada:

(1, 2, 3, 4, 5)

(-1, -2, -3, -4, -5)

Utilice zip para convertir la siguiente matriz 5x4 en una matriz 4x5

In [20]:
matriz5x4 = (( 1, 2, 3, 4),
             ( 5, 6, 7, 8),
             ( 9,10,11,12),
             (13,14,15,16),
             (17,18,19,20))

matriz4x5 = tuple(zip(*matriz5x4))




print(matriz4x5) #note que la respuesta es un tuple nuevamente

((1, 5, 9, 13, 17), (2, 6, 10, 14, 18), (3, 7, 11, 15, 19), (4, 8, 12, 16, 20))


Respuesta esperada:

((1, 5, 9, 13, 17), (2, 6, 10, 14, 18), (3, 7, 11, 15, 19), (4, 8, 12, 16, 20))

Utilice la función enumerate para adjuntar al final de cada uno de las listas cada palabra

Por ejemplo: 
- a = ['H. Montana', 'E. Hurley']
- b = [40, 60]
- Resultado: 'H. Montana 40' y así sucesivamente

In [22]:
twilight = ['Bella Swan', 'Edward Cullen', 'Jacob Black', 'Charlie Swan',
            'Carlisle Cullen', 'Esme Cullen', 'Alice Cullen']

for A, nombres in enumerate(twilight):
    print(f'{nombres} {A},')



Bella Swan 0,
Edward Cullen 1,
Jacob Black 2,
Charlie Swan 3,
Carlisle Cullen 4,
Esme Cullen 5,
Alice Cullen 6,


Resultado esperado:

('Bella Swan 0', 'Edward Cullen 1', 'Jacob Black 2', 'Charlie Swan 3', 'Carlisle Cullen 4', 'Esme Cullen 5', 'Alice Cullen 6')

### Listas de Comprensión

En python se pueden crear listas rápidas y precisas por medio e listas de comprensión.  Lo que creamos anterioremnte con for loop, por ejemplo, el codigo inferior...

In [167]:
algunas_provincias = ['PanAma', 'ColOn', 'DArien', 'VerAguas', 'ChIriqui', 'HErrera', 'LOs SAntOs']
en_mayusculas = []
for provincia in algunas_provincias:
    en_mayusculas.append(provincia.upper())

print(en_mayusculas)

['PANAMA', 'COLON', 'DARIEN', 'VERAGUAS', 'CHIRIQUI', 'HERRERA', 'LOS SANTOS']


... lo podemos hacer de una forma más sencilla y rápida utilizando listas de comprensión.  Veamos el ejemplo

In [164]:
en_mayusculas = [provincia.upper() for provincia in algunas_provincias]
print(en_mayusculas)

['PANAMA', 'COLON', 'DARIEN', 'VERAGUAS', 'CHIRIQUI', 'HERRERA', 'LOS SANTOS']


Note que de los dos ejemplos, en el último no necesitamos crear el arreglo y solo pasar a la variable el resultado.  Veamos más del siguiente ejemplo, creando una lista de valores al cubo

In [168]:
# con for loop
x = []
for i in range(10):
    x.append(i**3)
print(x)
    
# con listas de comprensión
x = [i**3 for i in range(10)]
print(x)

[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]
[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]


Tambien podemos añadir sentencias if a una lista de comprensión.

In [169]:
# solo listar de los numeros que son pares su valor al cubo
x = [i**3 for i in range(10) if i%2 == 0]
print(x)

[0, 8, 64, 216, 512]


Si queremos hacer sentencias con if/else entonces tenemos que mover la lista de comprensión al inicio

In [171]:
x = [i**3 if i%2 == 0 else i for i in range(10)]
print(x)

[0, 1, 8, 3, 64, 5, 216, 7, 512, 9]


#### Ahora ud.

De la siguiente lista de conquistadores solamente extraiga el primer nombre.

In [243]:
conquistadores = ['Vasco Núnez de Balboa', 'Luis de Moscoso Alvarado', 'Juan de Cárdenas',
                  'Pedro de Heredia', 'Gutierre de Miranda', 'Juan de Salcedo']
CQ=[name.split()[0] for name in conquistadores]
c2=str(c)[1:-1]
print(c2)



'Vasco', 'Luis', 'Juan', 'Pedro', 'Gutierre', 'Juan'


Respuesta esperada:

['Vasco', 'Luis', 'Juan', 'Pedro', 'Gutierre', 'Juan']

Calcule los números primos hasta el número 10 con la [fórmula de Fermat](https://es.wikipedia.org/wiki/N%C3%BAmero_de_Fermat) usando listas de comprensión.

In [249]:
prime=[2**2**n + 1 for n in range(0,10)]
print(prime)

[3, 5, 17, 257, 65537, 4294967297, 18446744073709551617, 340282366920938463463374607431768211457, 115792089237316195423570985008687907853269984665640564039457584007913129639937, 13407807929942597099574024998205846127479365820592393377723561443721764030073546976801874298166903427690031858186486050853753882811946569946433649006084097]


Resultado esperado:

[3, 5, 17, 257, 65537, 4294967297, 18446744073709551617, 340282366920938463463374607431768211457, 115792089237316195423570985008687907853269984665640564039457584007913129639937, 13407807929942597099574024998205846127479365820592393377723561443721764030073546976801874298166903427690031858186486050853753882811946569946433649006084097]

Con listas de comprensión intente hacer un filtro de los siguientes bateadores entre los años 1980 hasta el 2015.

In [308]:
major_league_players = {'Mel Allen':1978, 'Red Barber':1978, 'Jerry Coleman':2005,
          'Joe Garagiola, Sr.':1991, 'Curt Gowdy':1984, 'Ken Harrelson':2020,
          'Al Helfer':2019, 'Russ Hodges':1980}
newplays={ key:value for (key,value) in major_league_players.items() if 1980<= value<=2015 }
impresion=str(list(newplays.keys()))[1:-1]
print(impresion)

'Jerry Coleman', 'Joe Garagiola, Sr.', 'Curt Gowdy', 'Russ Hodges'


Respuesta esperada:
    
['Jerry Coleman', 'Joe Garagiola, Sr.', 'Curt Gowdy', 'Russ Hodges']